### PR TITLE
LVRUBYM-325: Implement I18n For application

### PR DESCRIPTION
**Pull request by skrypnyk88**
## JIRA

* [Main JIRA ticket](http://ssu-jira.softserveinc.com/browse/LVRUBYM-325)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-326)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-327)

## Code reviewers

- [x] @HelenMaherovska
- [x] @yaremakleputs
- [x] @Mykhailo123

### Second Level Review

- [ ] @omaik

## Summary of issue

Need implement I18n for application and add Teacher controller

## Summary of change

- added I18n for application and Teacher controller

## Testing approach

We do not need any tests here

## CHECK LIST
- [x]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: app/controllers/v1/teachers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      if current_user.update(teacher_params)
+        head :no_content
+      else
+        render json: { errors: current_user.errors.full_messages }
+      end
+    end
+
+    def teacher_params
+      params.require(:teacher)
+            .permit(:first_name, :last_name, :email, :phone, :locale, :id)
```
Comment (Mykhailo123) [inline app/controllers/v1/teachers_controller.rb]: Teacher can only edit locale

File: app/models/user.rb [hunk 2]
```diff
@@ -5,6 +5,7 @@ class User < ApplicationRecord
   devise :database_authenticatable, :validatable
   VALID_NAME_REGEX = /\A[a-zA-Z'-]+\z/
   VALID_PHONE_REGEX = /\A[\d\- ]{7,10}\z/
+  VALID_EMAIL_REGEX = /\A[\w+\-.']+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (Mykhailo123) [inline app/models/user.rb]: device has its own email-validation

File: app/controllers/v1/teachers_controller.rb (hunk 1 above)
Comment (Mykhailo123) [inline app/controllers/v1/teachers_controller.rb]: Teacher can only edit locale

File: app/models/user.rb (hunk 2 above)
Comment (Mykhailo123) [inline app/models/user.rb]: device has its own email-validation

File: app/controllers/v1/teachers_controller.rb (hunk 1 above)
Comment (Mykhailo123) [inline app/controllers/v1/teachers_controller.rb]: here http://app.letsshareapps.com/profiles/3/edit

Comment (skrypnyk88) [inline app/controllers/v1/teachers_controller.rb]: ок)

File: app/models/user.rb (hunk 2 above)
Comment (skrypnyk88) [inline app/models/user.rb]: But devise allow wrong email like "some@gmd"

File: .rubocop.yml [hunk 3]
```diff
@@ -5,6 +5,7 @@ AllCops:
   Exclude:
     - 'app/views/**/*'
     - 'db/schema.rb'
+    - 'db/seeds.rb'
```
Comment (omaik) [inline .rubocop.yml]: What is wrong with seeds?

File: app/models/user.rb (hunk 2 above)
Comment (omaik) [inline app/models/user.rb]: Then put validation to devise initializer

File: spec/factories/teachers.rb [hunk 4]
```diff
@@ -7,5 +7,6 @@
     password_confirmation 'password'
     gender { %w(male female other).sample }
     phone '0673355891'
+    locale { %w(en ua).sample }
   end
 end
```
Comment (omaik) [inline spec/factories/teachers.rb]: Why do you have the same factory for user and teacher?

File: .rubocop.yml (hunk 3 above)
Comment (skrypnyk88) [inline .rubocop.yml]: There are too many lines in block. Now 27 instead 25. When I will add my report will be 35 + other reports. I can duplicate that block, but then will be some question about DRY.

File: app/models/user.rb (hunk 2 above)
Comment (skrypnyk88) [inline app/models/user.rb]: Ok

File: spec/factories/teachers.rb (hunk 4 above)
Comment (skrypnyk88) [inline spec/factories/teachers.rb]: This is not my code. If I will delete this code 19 spec tests will fall. Should I need fix it?

File: .rubocop.yml (hunk 3 above)
Comment (omaik) [inline .rubocop.yml]: Then you can exclude it from block length check, not from all cops

File: spec/factories/teachers.rb (hunk 4 above)
Comment (omaik) [inline spec/factories/teachers.rb]: Yeah, I think it is better then hold to same files

File: .rubocop.yml (hunk 3 above)
Comment (skrypnyk88) [inline .rubocop.yml]: Ок.

File: app/controllers/authentication_controller.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 class AuthenticationController < ApplicationController
   skip_before_action :authenticate_user!, only: :authenticate
-
+  skip_before_action :set_locale
```
Comment (omaik) [inline app/controllers/authentication_controller.rb]: Put empty line between callbacks and methods

File: app/models/user.rb [hunk 6]
```diff
@@ -15,4 +15,5 @@ class User < ApplicationRecord
             length: { in: 7..10 },
             format: { with: VALID_PHONE_REGEX, message: 'use only digits' },
             allow_nil: true
+  enum locale: { en: 0, ua: 1 }
```
Comment (omaik) [inline app/models/user.rb]: Please do not put related parts throughout all the file. Hold all callbacks, validations, associations, enums etc together. And there with others :)

File: spec/controllers/v1/my_day_reports_controller_spec.rb [hunk 7]
```diff
@@ -9,6 +9,7 @@
 
   before do
     allow(subject).to receive(:authenticate_user!)
+    allow(controller).to receive(:set_locale)
```
Comment (omaik) [inline spec/controllers/v1/my_day_reports_controller_spec.rb]: Any problem with this method?

File: app/controllers/authentication_controller.rb (hunk 5 above)
Comment (skrypnyk88) [inline app/controllers/authentication_controller.rb]: Done

File: app/models/user.rb (hunk 6 above)
Comment (skrypnyk88) [inline app/models/user.rb]: Ok

File: spec/controllers/v1/my_day_reports_controller_spec.rb (hunk 7 above)
Comment (skrypnyk88) [inline spec/controllers/v1/my_day_reports_controller_spec.rb]: yes. I have an error "undefined local variable or method `current_user'". So I allowed to receive this method. Is this wrong?

Comment (omaik) [inline spec/controllers/v1/my_day_reports_controller_spec.rb]: Then talk to Andriy to remove this after he will set up authentication

Comment (skrypnyk88) [inline spec/controllers/v1/my_day_reports_controller_spec.rb]: He will rewrite his own tests. And Ok. I will tell him.

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -1,3 +1,15 @@
...
   before_action :authenticate_user!
+  before_action :set_locale
+
+  def set_locale
+    I18n.locale = current_user.locale || I18n.default_locale
+  end
+
+  private
+
+  def find_group
```
Comment (omaik) [inline app/controllers/application_controller.rb]: Do we need this?

File: app/controllers/v1/teachers_controller.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
+module V1
+  class TeachersController < ApplicationController
+    def show
+      @teacher = current_user
```
Comment (omaik) [inline app/controllers/v1/teachers_controller.rb]: Work always with @teacher or with current_user. If you mix them it is really confusing.

File: config/locales/en.yml [hunk 10]
```diff
@@ -1,23 +1,217 @@
...
-#
-# This would use the information in config/locales/es.yml.
-#
-# To learn more, please read the Rails Internationalization guide
-# available at http://guides.rubyonrails.org/i18n.html.
-
+---
 en:
-  hello: "Hello world"
+  unique_errors:
```
Comment (omaik) [inline config/locales/en.yml]: Better sort errors based on model they belong to

File: spec/controllers/v1/my_day_reports_controller_spec.rb (hunk 7 above)
Comment (omaik) [inline spec/controllers/v1/my_day_reports_controller_spec.rb]: no no no, better stub `curent_user` :) It is much more explicit and solves more problems

File: app/controllers/v1/teachers_controller.rb (hunk 9 above)
Comment (skrypnyk88) [inline app/controllers/v1/teachers_controller.rb]: Ok. Done.

**Comment by JulkaIII** [comment]
Add translation for devise messages for email and password fields

File: app/controllers/v1/teachers_controller.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    end
+
+    def update
+      if current_user.update(current_user_params)
+        head :no_content
+      else
+        render json: { errors: current_user.errors.full_messages }, status: 400
```
Comment (omaik) [inline app/controllers/v1/teachers_controller.rb]: use symbol notation of status

File: app/controllers/v1/students_controller.rb [hunk 12]
```diff
@@ -30,7 +31,11 @@ def find_student
     end
 
     def render_show_or_bad_request(succeed)
-      succeed ? (render :show) : (head :bad_request)
+      if succeed
+        render :show
+      else
+        render json: { errors: @student.errors.full_messages }, status: 400
```
Comment (omaik) [inline app/controllers/v1/students_controller.rb]: same here

File: app/controllers/v1/students_controller.rb [hunk 13]
```diff
@@ -1,6 +1,7 @@
 module V1
   class StudentsController < ApplicationController
     before_action :find_student, only: [:show, :update, :destroy]
+    skip_before_action :set_locale
```
Comment (omaik) [inline app/controllers/v1/students_controller.rb]: Any reason for this?

File: app/controllers/authentication_controller.rb [hunk 14]
```diff
@@ -1,5 +1,6 @@
 class AuthenticationController < ApplicationController
   skip_before_action :authenticate_user!, only: :authenticate
+  skip_before_action :set_locale
```
Comment (okohuch) [inline app/controllers/authentication_controller.rb]: Why are you doing this?

File: app/controllers/v1/teachers_controller.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+      end
+    end
+
+    def current_user_params
```
Comment (okohuch) [inline app/controllers/v1/teachers_controller.rb]: Could you rename the method appropriately? teacher_params??? This is not UsersController

File: app/controllers/authentication_controller.rb (hunk 14 above)
Comment (skrypnyk88) [inline app/controllers/authentication_controller.rb]: Current_user is not defined before authentication. So I need skip this method here

Comment (okohuch) [inline app/controllers/authentication_controller.rb]: @skrypnyk88 Ok, but you have 
`I18n.locale = current_user.locale || I18n.default_locale`

if no current_user, then the second condition is working

Comment (okohuch) [inline app/controllers/authentication_controller.rb]: maybe you get an error that there is no method `locale` for `Nil` class ???

If so, you can write like this:
`I18n.locale = current_user&.locale || I18n.default_locale` and then you will not have the error

File: app/controllers/v1/teachers_controller.rb (hunk 15 above)
Comment (skrypnyk88) [inline app/controllers/v1/teachers_controller.rb]: Yes I understand, teacher is Current_user in my case. So I defined current_user, but not user.

Comment (okohuch) [inline app/controllers/v1/teachers_controller.rb]: anyway, it's a bad name of the method!! It's not Users controller

File: app/controllers/authentication_controller.rb (hunk 14 above)
Comment (okohuch) [inline app/controllers/authentication_controller.rb]: http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/

File: app/controllers/v1/students_controller.rb (hunk 12 above)
Comment (skrypnyk88) [inline app/controllers/v1/students_controller.rb]: Done

File: app/controllers/v1/teachers_controller.rb (hunk 11 above)
Comment (skrypnyk88) [inline app/controllers/v1/teachers_controller.rb]: Done

File: app/controllers/authentication_controller.rb (hunk 14 above)
Comment (skrypnyk88) [inline app/controllers/authentication_controller.rb]: Cool. Thank you. :)

File: app/controllers/v1/teachers_controller.rb (hunk 15 above)
Comment (skrypnyk88) [inline app/controllers/v1/teachers_controller.rb]: Ok. I will fix if.

File: app/controllers/authentication_controller.rb [hunk 16]
```diff
@@ -4,20 +4,23 @@ class AuthenticationController < ApplicationController
...
+    send_response(current_user)
   end
 
   private
 
-  def send_token_header(user)
-    head :ok, auth_token: JsonWebToken.encode(user)
+  def send_response(user)
+    response.headers['Auth-Token'] = JsonWebToken.encode(user)
+    render json: user.attributes
```
Comment (omaik) [inline app/controllers/authentication_controller.rb]: Better use jbuilder for this

File: app/controllers/v1/teachers_controller.rb (hunk 9 above)
Comment (omaik) [inline app/controllers/v1/teachers_controller.rb]: Not done, actually.

Comment (omaik) [inline app/controllers/v1/teachers_controller.rb]: in teachers controller, you have `@teacher` `current_user` and `current_teacher_params` three different namings. Which is really confusing

File: app/controllers/authentication_controller.rb (hunk 14 above)
Comment (skrypnyk88) [inline app/controllers/authentication_controller.rb]: But there is another problem. A copy of Devise::Strategies::JWT has been removed from the module tree but is still active!

Comment (omaik) [inline app/controllers/authentication_controller.rb]: Then, try to fix it, assume that problem with constants autoloading.

File: app/controllers/authentication_controller.rb [hunk 17]
```diff
@@ -1,23 +1,26 @@
...
   end
 
   private
 
-  def send_token_header(user)
-    head :ok, auth_token: JsonWebToken.encode(user)
+  def send_response(user)
+    response.headers['Auth-Token'] = JsonWebToken.encode(user)
+    @current_user = user
+    render 'v1/users/show', status: :ok, formats: :json
```
Comment (omaik) [inline app/controllers/authentication_controller.rb]: any need for `formats` key?

File: app/controllers/v1/teachers_controller.rb [hunk 18]
```diff
@@ -0,0 +1,25 @@
+module V1
+  class TeachersController < ApplicationController
+    def show
+      @current_teacher = current_user
+    end
+
+    def update
+      @current_teacher = current_user
```
Comment (omaik) [inline app/controllers/v1/teachers_controller.rb]: Lines 4 and 8 totally the same, better use callback for it

Comment (omaik) [inline app/controllers/v1/teachers_controller.rb]: btw `@teacher` would be enough

Comment (skrypnyk88) [inline app/controllers/v1/teachers_controller.rb]: ок

File: app/controllers/authentication_controller.rb (hunk 17 above)
Comment (skrypnyk88) [inline app/controllers/authentication_controller.rb]: Yes. I need render my page as `json`.

Comment (omaik) [inline app/controllers/authentication_controller.rb]: Rails will do that automatically, if you are doing json request

Comment (skrypnyk88) [inline app/controllers/authentication_controller.rb]: ок

File: app/controllers/v1/teachers_controller.rb [hunk 19]
```diff
@@ -0,0 +1,30 @@
+module V1
+  class TeachersController < ApplicationController
+    before_action :find_teacher
+
+    def show; end
+
+    def update
+      if @teacher.update(teacher_params)
```
Comment (omaik) [inline app/controllers/v1/teachers_controller.rb]: `update_attributes`

**Comment by omaik** [comment]
:+1:

**Pull request merged**